File: behavioral/visitor/code_analyzer/python/code_analyzer/pretty_print_visitor.py
from .visitor import Visitor
from .ast_nodes import FunctionDefinitionNode, VariableDeclarationNode, AssignmentStatementNode, IfStatementNode, ExpressionStatementNode, ExpressionNode, Node

class PrettyPrintVisitor(Visitor):
    """Prints a textual representation of the AST with indentation, controlling traversal."""
    def __init__(self, indent_char="  "):
        self._output = ""
        self._indent_char = indent_char
        self._indent_level = 0

    def _indent(self):
        return self._indent_char * self._indent_level

    def reset(self):
        self._output = ""
        self._indent_level = 0

    def get_output(self):
        return self._output.strip() # Remove any trailing newline

    def visit_function_definition(self, element: FunctionDefinitionNode):
        params = ", ".join(element.parameters)
        self._output += f"{self._indent()}Function: {element.name}({params})\n"
        self._indent_level += 1
        for stmt in element.body:
            stmt.accept(self) # Visitor explicitly traverses children
        self._indent_level -= 1

    def visit_variable_declaration(self, element: VariableDeclarationNode):
        init_str = ""
        if element.initializer:
            # If initializer is complex, it might need its own accept call
            # For simple ExpressionNode, just accessing representation is ok
             init_str = f" = {element.initializer.representation}"
            # If initializer could be another statement/complex expr:
            # element.initializer.accept(self) # Would need careful output structuring
        type_hint = f": {element.type_hint}" if element.type_hint else ""
        self._output += f"{self._indent()}Declare: {element.name}{type_hint}{init_str}\n"

    def visit_assignment_statement(self, element: AssignmentStatementNode):
        # Similar to declaration, handle value access/visit
        value_repr = element.value.representation
        # If value could be complex:
        # element.value.accept(self)
        self._output += f"{self._indent()}Assign: {element.target_variable} = {value_repr}\n"

    def visit_if_statement(self, element: IfStatementNode):
        # Visit condition (if needed, e.g., if it's not just a simple string)
        # element.condition.accept(self)
        self._output += f"{self._indent()}If ({element.condition.representation}):\n"
        self._indent_level += 1
        for stmt in element.then_branch:
            stmt.accept(self) # Visitor explicitly traverses children
        self._indent_level -= 1

        if element.else_branch:
            self._output += f"{self._indent()}Else:\n"
            self._indent_level += 1
            for stmt in element.else_branch:
                stmt.accept(self) # Visitor explicitly traverses children
            self._indent_level -= 1

    def visit_expression_statement(self, element: ExpressionStatementNode):
        # Visit expression
        # element.expression.accept(self)
        self._output += f"{self._indent()}ExprStmt: {element.expression.representation}\n"

File: behavioral/visitor/code_analyzer/typescript/code_analyzer/pretty_print_visitor.ts
import { Visitor } from "./visitor";
import {
  FunctionDefinitionNode,
  VariableDeclarationNode,
  AssignmentStatementNode,
  IfStatementNode,
  ExpressionStatementNode,
} from "./ast_nodes";

// Prints the AST nodes in a structured, indented format.
export class PrettyPrintVisitor implements Visitor {
  private indentLevel: number = 0;
  private output: string = "";
  private readonly indentChar: string;

  constructor(indentChar: string = "  ") {
    this.indentChar = indentChar;
  }

  private indent(): string {
    return this.indentChar.repeat(this.indentLevel);
  }

  getOutput(): string {
    return this.output;
  }

  visitFunctionDefinition(element: FunctionDefinitionNode): void {
    this.output += `${this.indent()}Function: ${element.name}(${element.parameters.join(", ")})\n`;
    this.indentLevel++;
    // Explicitly traverse body with indentation
    element.body.forEach((stmt) => stmt.accept(this));
    this.indentLevel--; // Decrease indent after visiting body
  }

  visitVariableDeclaration(element: VariableDeclarationNode): void {
    const initStr = element.initializer
      ? ` = ${element.initializer.representation}`
      : "";
    this.output += `${this.indent()}Declare: ${element.name}: ${element.typeHint}${initStr}\n`;
    // Initializer traversal via its own accept
    element.initializer?.accept(this);
  }

  visitAssignmentStatement(element: AssignmentStatementNode): void {
    this.output += `${this.indent()}Assign: ${element.targetVariable} = ${element.value.representation}\n`;
    // Value traversal via its own accept
    element.value.accept(this);
  }

  visitIfStatement(element: IfStatementNode): void {
    this.output += `${this.indent()}If (${element.condition.representation}):\n`;
    // Visit condition first (does not usually affect indent)
    // element.condition.accept(this); // Typically not visited by pretty printer unless complex

    this.indentLevel++;
    // Visit 'then' branch
    element.thenBranch.forEach((stmt) => stmt.accept(this));
    this.indentLevel--;

    if (element.elseBranch) {
      this.output += `${this.indent()}Else:\n`;
      this.indentLevel++;
      // Visit 'else' branch
      element.elseBranch.forEach((stmt) => stmt.accept(this));
      this.indentLevel--;
    }
  }

  visitExpressionStatement(element: ExpressionStatementNode): void {
    this.output += `${this.indent()}ExprStmt: ${element.expression.representation}\n`;
    // Expression traversal via its own accept
    element.expression.accept(this);
  }

  // Reset indent level manually if needed between visiting different top-level nodes
  resetIndent(): void {
    this.indentLevel = 0;
  }
}
